### ISSUE TITLE
[HOLD for payment 2022-11-24] [$500] Highlighted emoji not visible after scrolling down with ArrowDown key after sometime - reported by @sobitneupane

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#8459]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open Emoji Picker.
2. Press down arrow multiple times to reach other categories of emoji (somewhere around "Animals & Nature).

## Expected Result:
Highlighted emoji is visible in the screen.

## Actual Result:
Highlighted emoji is not visible in the screen.

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Web

**Version Number:** 1.1.49-0
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/161472213-3a70ab2d-1cdd-46a0-ba37-9d5b4bb9e004.mov

**Upwork URL:** https://www.upwork.com/jobs/~0104371e92e24f551a
**Issue reported by:** @sobitneupane
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1648127603575109

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by chiragsalian** [comment]
lol funny bug, should be pretty straightforward to fix externally.

**Comment by trjExpensify** [comment]
Issue on upwork here: https://www.upwork.com/jobs/~01ad54324b48c3dfa9

**Comment by trjExpensify** [comment]
👋 back from OoO. No proposals on Upwork yet, so I've doubled [the job](https://www.upwork.com/jobs/~01ad54324b48c3dfa9) to $500. 

@sobitneupane as you reported this bug, are you interested in submitting a proposal to fix it?

**Comment by metehanozyurt** [comment]
I guess I'm not the priority, but I have a proposal.

### Cause 
Constant variable value is wrong. That's why zooming in the wrong place.

### Solution 
calculation is true but constant variable of EMOJI_PICKER_ITEM_HEIGHT(40) is wrong in this file 
https://github.com/Expensify/App/blob/ad637d3dd7dbb03cfdc6ebcb9196944c4f6165bb/src/CONST.js#L417

this should be change like this.

```js
EMOJI_PICKER_ITEM_HEIGHT: 34.5,
```
![WhatsApp Image 2022-04-24 at 17 22 31](https://user-images.githubusercontent.com/6093207/164981384-a65f85ec-b7b2-4444-8b4c-286f3cd367fe.jpeg)

https://user-images.githubusercontent.com/6093207/164981501-5198b5bd-cf66-4237-86a8-44bb95792973.mp4

**Comment by parasharrajat** [comment]
Good catch @metehanozyurt but do you mind explaining how is that causing this issue?

**Comment by metehanozyurt** [comment]
Thank you @parasharrajat 🙏 . 
When we click arrows this functions calls (`highlightAdjacentEmoji`). Then comes switch for index calculation.

https://github.com/Expensify/App/blob/829595e9e2e24b6637d5c89bd553f4f0623c91b8/src/components/EmojiPicker/EmojiPickerMenu/index.js#L265-L271
and
https://github.com/Expensify/App/blob/829595e9e2e24b6637d5c89bd553f4f0623c91b8/src/components/EmojiPicker/EmojiPickerMenu/index.js#L284-L294

If highlightedIndex index changed then calls `scrollToHighlightedIndex ` function

https://github.com/Expensify/App/blob/829595e9e2e24b6637d5c89bd553f4f0623c91b8/src/components/EmojiPicker/EmojiPickerMenu/index.js#L300-L303

`scrollToHighlightedIndex` function calculate for how much scroll the emoji list. When we press the down button, the emoji list starts to scroll extra pixel. Because the height we give as a parameter is 40 pixels but every line of emoji list have 34.5 pixel height. Thats why we loose highlighted emoji .😇

**Comment by sobitneupane** [comment]
**Proposal:**
Constant defined in CONST.js seems incorrect.
Replace following lines of code by:
https://github.com/Expensify/App/blob/de6404b06c6d886b9495135aab0d8c1f5161a6b3/src/CONST.js#L416-L417

```
    NON_NATIVE_EMOJI_PICKER_LIST_HEIGHT: 295.5,
    EMOJI_PICKER_ITEM_HEIGHT: 34.5,
```
Above change will not be enough to solve the issue.
Calculation of numEmojiRows is incorrect. Here value of `this.state.highlightedIndex` is not correct. The reason is addition of unnecessary space from `getDynamicSpacing()` function. If checked, we can see that the index of very first emoji in the list is 16 which is supposed to be 8.
 https://github.com/Expensify/App/blob/de6404b06c6d886b9495135aab0d8c1f5161a6b3/src/components/EmojiPicker/EmojiPickerMenu/index.js#L320

https://github.com/Expensify/App/blob/de6404b06c6d886b9495135aab0d8c1f5161a6b3/src/libs/EmojiUtils.js#L95-L106
```diff
function getDynamicSpacing(emojiCount, suffix) {
    const spacerEmojis = [];
    let modLength = CONST.EMOJI_NUM_PER_ROW - (emojiCount % CONST.EMOJI_NUM_PER_ROW);
-    while (modLength > 0) {
+    while (modLength > 0 && modLength <  CONST.EMOJI_NUM_PER_ROW) {
        spacerEmojis.push({
            code: `${CONST.EMOJI_SPACER}_${suffix}_${modLength}`,
            spacer: true,
        });
        modLength -= 1;
    }
    return spacerEmojis;
}
```

**Comment by parasharrajat** [comment]
Ok, I see. Both of the proposals are suggesting to change the static dimensions as the first step. But instead of doing that How will we make sure that rendered emojis have 40px height. We are expecting the emoji to take 40px height. 

@sobitneupane I will need more information to understand your proposal. What are you doing in `getDynamicSpacing` and why? 

> Value of numEmojiRows is incorrect. Here value of this.state.highlightedIndex is not correct. The reason is addition of unnecessary space from getDynamicSpacing() function. If checked, we can see that the index of very first emoji in the list is 16 which is supposed to be 8.

Didn't get it.

**Comment by metehanozyurt** [comment]
It's used to be 40, but it has been changed. For this reason, 40 appears to be fixed values.
This commit add 40
https://github.com/Expensify/App/commit/fe374e3d2a3f034c7f68a00a066332ef4891cf9f

this commit changes this value
https://github.com/Expensify/App/commit/2a6e5b9323e2e8c6bd423a0c3d67169fe4415238

**Comment by parasharrajat** [comment]
Ok, I see.

**Comment by sobitneupane** [comment]
> @sobitneupane I will need more information to understand your proposal. What are you doing in `getDynamicSpacing` and why?
> 
> > Value of numEmojiRows is incorrect. Here value of this.state.highlightedIndex is not correct. The reason is addition of unnecessary space from getDynamicSpacing() function. If checked, we can see that the index of very first emoji in the list is 16 which is supposed to be 8.

As per my understanding `getDynamicSpacing` is supposed to add spaces to fill emoji places at the end of categories with spaces if there are less than 8 emojis in that row (and seven spaces for header row). For example, if there are 5 emojis in the last row, 3 spaces are added. But `getDynamicSpacing` is adding 8 spaces even if the last row is full with 8 emojis as `modLength = CONST.EMOJI_NUM_PER_ROW - (emojiCount % CONST.EMOJI_NUM_PER_ROW);` is greater than 0 in this case as well. So, adding `modLength <  CONST.EMOJI_NUM_PER_ROW` in `while` loop condition will prevent to do so.

This issue is effecting the calculation of  `numEmojiRows`.

**Comment by trjExpensify** [comment]
@parasharrajat any feedback on the above?

**Comment by parasharrajat** [comment]
Ok. @sobitneupane analysis is correct. It took me some time to figure out the `getDynamicSpacing` 's role. Maybe add a few comments in the PR. Apart from it, we can use the full px values instead of floats. 

@sobitneupane's proposal looks good to me. 

cc: @chiragsalian 


:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by chiragsalian** [comment]
i think the proposal by @sobitneupane looks good to me as well. Feel free to create the PR @sobitneupane.

**Comment by trjExpensify** [comment]
Perfect! @sobitneupane, can you apply for the job here please? https://www.upwork.com/jobs/~01ad54324b48c3dfa9

**Comment by trjExpensify** [comment]
Looks like that job post expired, @sobitneupane. Created a new one here for you to apply to: https://www.upwork.com/jobs/~0104371e92e24f551a

**Comment by trjExpensify** [comment]
Back from OoO today, so I've hired you on Upwork @sobitneupane. 

Update for Melv: PR is in review, it's on @parasharrajat to review. 👍

**Comment by parasharrajat** [comment]
Any update @sobitneupane on the PR?

**Comment by sobitneupane** [comment]
> Any update @sobitneupane on the PR?

Updated the PR.

**Comment by parasharrajat** [comment]
I am trying to find another solution.

**Comment by trjExpensify** [comment]
Okay, is there a thread somewhere I can subscribe to in #expensify-open-source?

**Comment by parasharrajat** [comment]
No discussion yet.

**Comment by trjExpensify** [comment]
@parasharrajat what's going on with this issue? Can we get an update, please.

**Comment by trjExpensify** [comment]
@parasharrajat how'd it go [here](https://github.com/Expensify/App/issues/8459#issuecomment-1205514580) a couple of weeks ago?

**Comment by trjExpensify** [comment]
👋 - what's going on [with the PR](https://github.com/Expensify/App/pull/8902) in review, how can we unblock it?

**Comment by trjExpensify** [comment]
Thank you!

**Comment by trjExpensify** [comment]
Bump on the above, @parasharrajat?

**Comment by parasharrajat** [comment]
No, I will just do that in a few minutes.

**Comment by parasharrajat** [comment]
Started with this https://github.com/Expensify/App/pull/8902#issuecomment-1276734709. Let's see if we need to go to slack. I will wrap up this PR asap but it depends on the discussion.

**Comment by trjExpensify** [comment]
The blocker has been worked through, so @sobitneupane will you be able to get the necessary changes in place this week so we can put this back in review? 🙏

**Comment by sobitneupane** [comment]
> The blocker has been worked through, so @sobitneupane will you be able to get the necessary changes in place this week so we can put this back in review? 🙏

PR ready for review.

**Comment by trjExpensify** [comment]
Awesome, thanks!

**Comment by trjExpensify** [comment]
PR is back in review. @parasharrajat has a question for @shawnborton [here](https://github.com/Expensify/App/pull/8902#issuecomment-1295934753) re: the emoji list in the EmojiPicker.

**Comment by shawnborton** [comment]
Commented on the PR.

**Comment by trjExpensify** [comment]
Thanks, @shawnborton! @parasharrajat, looks like the PR is ready for you again to complete the testing, though I've asked @sobitneupane to confirm.

**Comment by parasharrajat** [comment]
Yeah, it's ready. Will be reviewed by tomorrow and probably merged.

**Comment by trjExpensify** [comment]
Don't get me excited for the Friday treats, @parasharrajat! 😅

**Comment by parasharrajat** [comment]
No treats for me. :rofl:

**Comment by trjExpensify** [comment]
👋 Did you get your Mac issues sorted out @parasharrajat?

**Comment by parasharrajat** [comment]
Yeah, I back on track.

**Comment by trjExpensify** [comment]
Good to hear. Let's get this one off the books!

**Comment by trjExpensify** [comment]
Where are we at here? It's almost Friday, again.. 😅

**Comment by parasharrajat** [comment]
> The PR that introduced the bug has been identified. Link to the PR:

Not sure which one. But I think this issue is since the start of EmojiPicker.

**Comment by trjExpensify** [comment]
Niceeee! Super great to see this one merged! ❤️ Let's get started on collaborating to check these off: 

> @parasharrajat @sobitneupane @chiragsalian The PR that introduced the bug has been identified. Link to the PR:

From [this comment](https://github.com/Expensify/App/issues/8459#issuecomment-1110072501), we reference [this commit](https://github.com/Expensify/App/commit/33b210ffe67785a60ee0ed0ffa2208b4d88d0a03) in [this PR](https://github.com/Expensify/App/pull/1991). That was a PR to add this new feature at the time, so is it accurate that this wasn't a regression, just missed in the initial implementation? 

>A discussion in #contributor-plus has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

Any suggestions on steps to take to avoid it happening again? Did this [code comment](https://github.com/Expensify/App/pull/8902/files#r1012160276) suggestion end up getting added to better explain what is happening, for example? 

>A [regression test](https://expensify.testrail.io/index.php?/dashboard) has been added or updated so that the same bug will not reach production again.

Happy to add this. I think [this existing test case](https://expensify.testrail.io/index.php?/cases/view/1971187&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=229062) can be updated with the following additions from #4: 

1. Navigate to a conversation
2. Open the emoji picker
3. Scroll down the list to the bottom verifying each emoji is rendered correctly
4. Close the emoji picker and reopen it again
5. Press the down arrow key multiple times to reach the "Animals & Nature" category
6. **Verify** the highlighted emoji is visible on the screen with every click. 
7. Press the up arrow key multiple times to reach the top row of the emojis
8. **Verify** the highlighted emoji is visible on the screen with every click. 

Thoughts?

**Comment by parasharrajat** [comment]
> From https://github.com/Expensify/App/issues/8459#issuecomment-1110072501, we reference [this commit](https://github.com/Expensify/App/commit/33b210ffe67785a60ee0ed0ffa2208b4d88d0a03) in https://github.com/Expensify/App/pull/1991. That was a PR to add this new feature at the time, so is it accurate that this wasn't a regression, just missed in the initial implementation?

My bad, not. I forgot that the thread was live.

**Comment by parasharrajat** [comment]
@sobitneupane Could you please create a small follow-up PR for adding code comments? https://github.com/Expensify/App/pull/8902/files#r1012160276 
No need to add screenshots etc. Just comments.

**Comment by trjExpensify** [comment]
>My bad, not. I forgot that the thread was live. Will handle this in follow-up.

^^ was that comment in reference to the below? 😅 

>Any suggestions on steps to take to avoid it happening again? Did this [code comment](https://github.com/Expensify/App/pull/8902/files#r1012160276) suggestion end up getting added to better explain what is happening, for example?

**Comment by parasharrajat** [comment]
Aha, :smile: my mind is running faster than my eyes and fingers.

**Comment by trjExpensify** [comment]
@sobitneupane did you create [this PR](https://github.com/Expensify/App/issues/8459#issuecomment-1316037517)? I can't find it.

**Comment by sobitneupane** [comment]
Oops, missed it. I will create one asap.

**Comment by trjExpensify** [comment]
Awesome, thanks!

**Comment by trjExpensify** [comment]
Alright, thanks for taking care of that. Closing! 🎉

**Comment by trjExpensify** [comment]
^^ Ignoring. Triggered by the quick follow-up PR to add a code comment.

**Comment by trjExpensify** [comment]
Can you send me a link to said contract?

**Comment by trjExpensify** [comment]
Awesome, thanks. Done!
